File: leet_code_algorithms/implement_queue_using_stacks/Implement_queue_using_stacks.java
package implement_queue_using_stacks;

import java.util.EmptyStackException;
import java.util.Stack;

class MyQueue {
    Stack<Integer> stack;

    public MyQueue() {
        this.stack = new Stack<Integer>();
    }

    public void push(int x) {
        this.stack.push(x);
    }

    public int pop() {
        if (this.stack.isEmpty()) {
            throw new EmptyStackException();
        }
        Stack<Integer> tempStack = new Stack<>();
        while (stack.size() > 1) {
            tempStack.push(stack.pop());
        }
        int firstNumber = stack.pop();
        while (!tempStack.isEmpty()) {
            stack.push(tempStack.pop());
        }
        return firstNumber;
    }

    public int peek() {
        return this.stack.get(0);
    }

    public boolean empty() {
        return this.stack.size() == 0;
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: leet_code_algorithms/implement_queue_using_stacks/implement_queue_using_stacks.ts
class MyQueue {
  stack: any[];
  constructor() {
    this.stack = [];
  }

  push(x: number): void {
    this.stack.push(x);
  }

  pop(): number {
    return this.stack.shift();
    //const [top,...rest] = this.queue
    //this.queue = rest
    //return top
  }

  peek(): number {
    return this.stack[0];
  }

  empty(): boolean {
    return this.stack.length === 0;
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
